### PR TITLE
Gzip images when deploying

### PR DESCRIPTION
**Pull request by pimterry**
Connects to #549

This gzips the Docker image stream before uploading it. https://github.com/resin-io/resin-builder/issues/187 doesn't actually seem to be required - I think express automatically decodes the body if you send it with the right headers. Works fine with no changes in my devenv and in prod, and it's significantly quicker than without gzipping (in prod, for a simple node hello world, a full `resin deploy --build` takes me 1m47 vs 2m40 on master).

The gzipping here is very easy - the tricky bit is providing a useful progress bar whilst doing so. I've gone for tracking the progress of draining the raw stream (that's piped into the gzip stream that's piped into the request), since it's quite a bit quicker than gzipping the whole buffer up front, which is the only easy way to get the gzipped size I can think of. If it's different to the actual request progress, it's imperceptibly so.

### CODE REVIEW

File: lib/actions/deploy.coffee [hunk 1]
```diff
@@ -186,21 +193,20 @@ module.exports =
 							{ image: imageName, log: '' }
 					.then ({ image: imageName, log: buildLogs }) ->
 						logs = buildLogs
-						Promise.join(
-							dockerUtils.bufferImage(docker, imageName, tmpPath)
+						Promise.all [
```
Comment (emirotin) [inline lib/actions/deploy.coffee]: Why switching here? `.join` is known t be more performant

File: lib/utils/docker.coffee [hunk 2]
```diff
@@ -281,29 +281,20 @@ exports.runBuild = (params, options, getBundleInfo, logStreams) ->
...
-			.pipe(stream)
-	.then ->
-		new Promise (resolve, reject) ->
-			fs.createReadStream(tmpFile)
-			.on 'open', ->
-				resolve(this)
-			.on('error', reject)
+	imageMetadata = image.inspectAsync()
+
+	Promise.all([image.get(), imageMetadata.get('Size')])
```
Comment (emirotin) [inline lib/utils/docker.coffee]: Same: `.join` can be used

File: lib/actions/deploy.coffee (hunk 1 above)
Comment (CameronDiver) [inline lib/actions/deploy.coffee]: I personally do find this slightly clearer to read here, I wasn't very keen on the `Promise.join` when I wrote it, this at least makes it obvious that the `performUpload` is the function.

File: lib/utils/docker.coffee (hunk 2 above)
Comment (CameronDiver) [inline lib/utils/docker.coffee]: Conversely I find this is a little less clear - not by much mind you - but I would probably use a `.join` here.

File: lib/actions/deploy.coffee (hunk 1 above)
Comment (pimterry) [inline lib/actions/deploy.coffee]: Ok, I'm going to leave this one

File: lib/utils/docker.coffee (hunk 2 above)
Comment (pimterry) [inline lib/utils/docker.coffee]: I've fixed this one

**Pull request merged**